### PR TITLE
k-quants with super-block size of 64

### LINKED ISSUE #1919: Make k-quants work with tensor dimensions that are not multiple of 256
As discussed in #1602, k-quants do not work for the Falcon-7B model. This is due to the fact that the number of columns in many tensors (`4544`) is not divisible by `256`, which is the super-block size of the k-quants.

It would be useful if k-quants could be adapted to work in such cases.

### PR DESCRIPTION
**Pull request by ikawrakow**
The main purpose of this PR is to enable k-quant usage for models where tensor sizes are not a multiple of the default k-quants super-block size of 256 such as OpenLLaMA-3B and Falcon-7B. This closes #1919 

The initial idea for handling this case was that we will simply pad tensor rows to be multiple of 256 and I initially started along these lines. But the change was turning much too big for my taste and was not just in isolated places related to the k-quants (as this PR), but was massively affecting the entire `ggml`. Hence, I eventually abandoned this approach.

Then there was the idea to use row-wise padding to 256 that but get away without major changes to the code  by simply using appropriate tensor views before operations that depend on the row size (e.g., rms norm). But that did not (easily) work because of the attention heads that make it necessary to have the padding in the middle to not change the embeddings each head is seeing, so this becomes too complicated as well.

So, at the end, here is the poor man's solution with an optional, compile-time configurable, super-block size of 64 for k-quants. To use it, simple
```
LLAMA_QKK_64=1 make -j
```
or
```
cmake -DLLAMA_QKK_64=1 ..
```
It is better to only enable `LLAMA_QKK_64` if you are working with models that need it. Using super-blocks of 64 instead of 256 a) loses most of the model size saving enabled by super-blocks of 256, and b) comes with a performance penalty in the 5-15% range (except on Metal, where super-blocks of 64 run 5-10% faster without me understanding why). In addition, `Q2_K` becomes completely useless as its size with super-blocks of 64 is basically the same as `Q3_K_S`. 

The PR provides implementation for CPU Scalar, ARM_NEON, AVX2, CUDA, Metal. OpenCL and AVX are missing.

In retrospect, it would have been better to finish the `ggml` modifications necessary for general support of tensor sizes not divisible by the quantization block size. I think this should be the longer-term goal. This will make `ggml` future proof against somebody coming up with the idea that tensor sizes should be picked, say, from the Fibonacci sequence rather than the currently more common approach of using powers of 2.  

Below are some graphs that show perplexity versus qunatized model size for OpenLLaMA-3B and LLaMA-7B.  The first graph is for OpenLLaMA-3B, and I have added the `Q4_0, Q4_1, Q5_0` and `Q5_1` values in red for comparison. The second graph is for LLaMA-7B. The black circles/lines are for the default super-block size of 256, the red for super-blocks of 64, and the `Q4_0...Q5_1` values are shown in blue.
 
ol3b_perplexity
ppl_vs_size_7B_a

### CODE REVIEW

File: k_quants.c [hunk 1]
```diff
@@ -1840,18 +2606,23 @@ void ggml_vec_dot_q5_K_q8_K(const int n, float * restrict s, const void * restri
...
         memcpy(utmp, x[i].scales, 12);
         utmp[3] = ((utmp[2] >> 4) & kmask2) | (((utmp[1] >> 6) & kmask3) << 4);
         const uint32_t uaux = utmp[1] & kmask1;
         utmp[1] = (utmp[2] & kmask2) | (((utmp[0] >> 6) & kmask3) << 4);
         utmp[2] = uaux;
         utmp[0] &= kmask1;
+#else
+        // TODO
+        const float d = 0, dmin = 0;
+#endif
```
Comment (ggerganov) [inline k_quants.c:2625]: Postponed for later or did you missed to implement this?

**Review by ggerganov — APPROVED** [review]
🦙

**Comment by Cebtenzzre** [comment]
I think you mean `LLAMA_QKK_64=1 make`

**Comment by ggerganov** [comment]
@Cebtenzzre fixed

**Comment by ikawrakow** [comment]
No, it is not possible to mix. It is a compile-time decision. This is why I was proposing the row-wise quantization in https://github.com/ggerganov/llama.cpp/discussions/5063. Given the lack of enthusiasm I saw there, it is not very likely I will add this to `llama.cpp`.

**Pull request merged**